research/delf/delf/python/datasets/tuples_dataset.py

```diff
@@ -0,0 +1,356 @@
+
+    Returns:
+      meta: String, containing TupleDataset meta.
+    """
+    fmt_str = self.__class__.__name__ + '\n'
+    fmt_str += '\tName and mode: {} {}\n'.format(self._name, self._mode)
+    fmt_str += '\tNumber of images: {}\n'.format(len(self.images))
+    fmt_str += '\tNumber of training tuples: {}\n'.format(len(self._qpool))
+    fmt_str += '\tNumber of negatives per tuple: {}\n'.format(self._nnum)
+    fmt_str += '\tNumber of tuples processed in an epoch: {}\n'.format(
+            self._qsize)
+    fmt_str += '\tPool size for negative remining: {}\n'.format(self._poolsize)
+    return fmt_str
+
+  def create_epoch_tuples(self, net):
+    """Creates epoch tuples with the hard-negative re-mining.
+
+    Negative examples are selected from clusters different than the cluster
+    of the query image, as the clusters are ideally non-overlaping. For
+    every query image we choose  hard-negatives, that is, non-matching images
+    with the most similar descriptor. Hard-negatives depend on the current
+    CNN parameters. K-nearest neighbors from all non-matching images are
+    selected. Query images are selected randomly. Positives examples are
+    fixed for the related query image during the whole training process.
+
+    Args:
+      net: Model, network to be used for negative re-mining.
+
+    Returns:
+      avg_l2: Float, average negative L2-distance.
+    """
+    self._n = 0
+    global_features_utils.debug_and_log(
+            '>> Creating tuples for an epoch of {}-{}...'.format(self._name,
+                                                                 self._mode),
+            True)
+    global_features_utils.debug_and_log(">> Used network: ", True)
+    global_features_utils.debug_and_log(net.meta_repr(), True)
+
+    ## Selecting queries.
+    # Draw `qsize` random queries for the tuples.
+    idx_list = np.arange(len(self._qpool))
+    np.random.shuffle(idx_list)
+    idxs2qpool = idx_list[:self._qsize]
+    self._qidxs = [self._qpool[i] for i in idxs2qpool]
+
+    ## Selecting positive pairs.
+    # Positives examples are fixed for each query during the whole training
+    # process.
+    self._pidxs = [self._ppool[i] for i in idxs2qpool]
+
+    ## Selecting negative pairs.
+    # If nnum = 0 create dummy nidxs.
+    # Useful when only positives used for training.
+    if self._nnum == 0:
+      self._nidxs = [[] for _ in range(len(self._qidxs))]
+      return 0
+
+    # Draw poolsize random images for pool of negatives images.
+    idx_list = np.arange(len(self.images))
+    np.random.shuffle(idx_list)
+    idxs2images = idx_list[:self._poolsize]
+
+    global_features_utils.debug_and_log(
+            '>> Extracting descriptors for query images...', debug=True)
+
+    img_list = self._img_names_to_full_path([self.images[i] for i in
+                                             self._qidxs])
+    qvecs = extract_descriptors_from_image_paths(
+            net,
+            image_paths=img_list,
```

Comment:
similarly to below, I think you can remove the `img_list` variable and simply inline the contents here.